src/lib/Remote.ts

```diff
@@ -0,0 +1,218 @@
+import { throttle } from "lodash";
+import { Messaging } from "./Messaging";
+import { LocalStorage } from "./LocalStorage";
+import { StorageKey } from "./Constants";
+import { Variable } from "../core/variables/Variable";
+
+/**
+ * The Remote class is a singleton class that provides the ability to store
+ * and retrieve Variables on a remote controller. It also provides listeners
+ * for Variable updates done remotely.
+ * @class
+ */
+export class Remote  {
+
+  /**
+   * Initializes a new instance of Remote.
+   * @private
+   * @static
+   * @return {Remote} A new instance of Remote.
+   */
+  private static _sharedInstance = new Remote();
+
+  /**
+   * Throttles network save calls.
+   * @private
+   * @static
+   * @type {any}
+   */
+  private static _throttle: any;
+
+  /**
+   * The remote ID.
+   * @private
+   * @type {string}
+   */
+  private remoteId: string;
+
+  /**
+   * Whether the remote controller is enabled.
+   * @private
+   * @type {string}
+   */
+  private enabled: boolean = false;
+
+  /**
+   * Initializes the remote controller.
+   *
+   * A call to this method will allow you to share your Variables to the
+   * remote controller being hosted as per your firebase configuration.
+   * @static
+   * @param {{}} config The firebase credentials.
+   */
+  static initializeRemote(config: {}): void {
+    // Get the locally stored remoteId. If doesn't exist, generate a new one
+    // and store it.
+    let storedRemoteId = this.getStoredRemoteId();
+    if (!storedRemoteId) {
+      storedRemoteId = this.storeRemoteId(this.generateRemoteId());
+    }
+    this._sharedInstance.remoteId = storedRemoteId;
+    firebase.initializeApp(config);
+  }
+
+  /**
+   * Returns the remote id from local storage.
+   * @private
+   * @static
+   * @return {string} Returns the remote id.
+   */
+  private static getStoredRemoteId(): string {
+    return LocalStorage.getPreference(StorageKey.KEY_REMOTE_ID);
+  }
+
+  /**
+   * Stores the remote id to local storage.
+   * @private
+   * @static
+   * @param  {string} remoteId The remote id.
+   * @return {string}          Returns the remote id.
+   */
+  private static storeRemoteId(remoteId: string): string {
+    LocalStorage.savePreference(StorageKey.KEY_REMOTE_ID, remoteId);
+    return remoteId;
+  }
+
+  /**
+   * Generates a unique id consisting of 8 chars.
+   * @private
+   * @static
+   * @return {string} Returns the new remote id.
+   */
+  private static generateRemoteId(): string {
+    return uuid().substring(0, 8);
+  }
+
+  /**
+   * Returns a database reference to the remixer instance.
+   * @private
+   * @return {firebase.database.Reference} The firebase database reference.
+   */
+  private dbReference(): firebase.database.Reference {
+    return firebase.database().ref(`remixer/${this.remoteId}`);
+  }
+
+  /**
+   * Start sharing the variable updates to the remote controller.
+   * @static
+   */
+  static startSharing(): void {
```

Comment:
I thought we were going to start sharing automatically as soon as you enable it the first time? that's probably another preference we need to save